### PR TITLE
Implement GetIndexedField for map-typed columns

### PR DESCRIPTION
**Pull request by swgillespie**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/7824.

## Rationale for this change

It's impossible to write a logical plan to query a column from a Parquet data source whose type is `Map`. The `Map` type is not explicitly supported by `GetIndexedField`. Maps are a useful column, are already supported by both Arrow and Parquet, and it makes sense to support it here. 

## What changes are included in this PR?

This commit extends the `NamedStructField` FieldAccess type to understand the `Map` data type. I chose this because the DataFusion SQL frontend parses the expression `x['y']` into a `NamedStructField`, which is a reasonable thing to do if we require that the argument to `x` be a constant scalar (which it is, in this implementation).

The Arrow Map array is essentially a list of structs, where each struct is a two-field struct. The first field of the struct is the key, and the second field of the struct is the value. Arrow traditionally names these `key` and `value`, though this implementation does not assume what they are named and instead assumes that the second column is the `value` column and the first is the `key` column, which is the same assumption made by the Arrow implementation we use.

To execute a mapped index access, we first scan the key column to identify entries that match the key that we are indexing, and again to gather the values corresponding to the keys that were selected.

## Are these changes tested?

This PR adds a new test, `map.slt`, which includes a Parquet file with two `Map` columns (one mapping strings to strings, the other mapping strings to ints) and writes some queries that use them.

## Are there any user-facing changes?

This change allows for the `GetIndexedField` type to now be usable with columns of type `Map`, which was not possible before.

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/get_indexed_field.rs [hunk 1]
```diff
@@ -183,6 +186,14 @@ impl PhysicalExpr for GetIndexedFieldExpr {
         let array = self.arg.evaluate(batch)?.into_array(batch.num_rows());
         match &self.field {
             GetFieldAccessExpr::NamedStructField{name} => match (array.data_type(), name) {
+                (DataType::Map(_, _), ScalarValue::Utf8(Some(k))) => {
+                    let map_array = as_map_array(array.as_ref())?;
+                    let key_scalar = Scalar::new(StringArray::from(vec![k.clone()]));
+                    let keys = arrow_ord::cmp::eq(&key_scalar, map_array.keys())?;
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/get_indexed_field.rs:192]: I think [`arrow::compute::eq`](https://docs.rs/arrow/latest/arrow/compute/kernels/cmp/fn.eq.html) is probably the more standard way to compare two arrays (though they call the same underlying kernels). That would let you avoid having to add the (newly explict) `arrow-ord` dependency.

**Review by alamb — APPROVED** [review]
Thank you @swgillespie - this looks very nice. 

I also played around with the file locally with `datafusion-cli` and it worked great. 

Thank you 🙏 

```sql
❯ describe 'parquet_map.parquet';
+-------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
| column_name | data_type                                                                                                                                                                                                                                                                                                                                              | is_nullable |
+-------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
| ints        | Map(Field { name: "entries", data_type: Struct([Field { name: "key", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "value", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }]), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, false) | NO          |
| strings     | Map(Field { name: "entries", data_type: Struct([Field { name: "key", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "value", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }]), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, false)  | NO          |
| timestamp   | Utf8                                                                                                                                                                                                                                                                                                                                                   | NO          |
+-------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
3 rows in set. Query took 0.018 seconds.
```

**Comment by alamb** [comment]
@swgillespie  I think you could avoid adding the arrow-ord dependency, but I don't think that is critical (as arrow-ord is already a transitive dependency via arrow anyways)

**Comment by swgillespie** [comment]
@alamb no problem - done!

File: datafusion/physical-expr/src/expressions/get_indexed_field.rs [hunk 2]
```diff
@@ -183,6 +186,14 @@ impl PhysicalExpr for GetIndexedFieldExpr {
         let array = self.arg.evaluate(batch)?.into_array(batch.num_rows());
         match &self.field {
             GetFieldAccessExpr::NamedStructField{name} => match (array.data_type(), name) {
+                (DataType::Map(_, _), ScalarValue::Utf8(Some(k))) => {
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/get_indexed_field.rs:189]: 👨‍🍳 👌

**Review by alamb — APPROVED** [review]
Really nice @swgillespie  -- thank you 🙏

**Pull request merged**